Fix crash of laser_assembler if service called before receiving scans

bobhenz-jabil opened the pull request:
Although this is extremely unlikely to hit on an actual robot, this code
would SEGFAULT if the service was called before any scans had been
received.

The code that crashes is a few lines down:
`scan_hist_.rbegin()->header.stamp`.

I ran into this when trying to run the tilt lidar nodes against a bag file.
In that case it is very easy to hit this problem.

This bug does not exist in newer branches of the repo since this code has
been rewritten, therefore this change does not need to be upstreamed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1)
<!-- Reviewable:end -->

c-andy-martin reviewed (changes requested):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1)* status: all files reviewed, 2 unresolved discussions (waiting on @bobhenz-jabil)

___
*[include/laser_assembler/base_assembler.h, line 293 at r1](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1#-MMRjkIOBhPun97c8D8w:-MMRjkIOBhPun97c8D8x:b5ploit) ([raw file](https://github.com/badgertechnologies/laser_assembler/blob/6c3063be938a8b1366e1c62b705028847693a945/include/laser_assembler/base_assembler.h#L293)):*
> ```objc
>   boost::unique_lock<boost::mutex> scan_hist_lock(scan_hist_mutex_) ;
> 
>   if (scan_hist_.size() == 0)
> ```

per slack convo, I would prefer instead to remove this and wait in the case we need to
___
*[include/laser_assembler/base_assembler.h, line 307 at r1](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1#-MMRjqJ1AQjM-LHLRK55:-MMRjqJ1AQjM-LHLRK56:bcc9q3) ([raw file](https://github.com/badgertechnologies/laser_assembler/blob/6c3063be938a8b1366e1c62b705028847693a945/include/laser_assembler/base_assembler.h#L307)):*
> ```objc
>     // scheduling delays and topic racing, sim time is not appropriate.
>     boost::chrono::steady_clock::time_point end_tp = boost::chrono::steady_clock::now() + boost::chrono::nanoseconds(blocking_period.toNSec());
>     while (boost::chrono::steady_clock::now() < end_tp && scan_hist_.rbegin()->header.stamp < req.end)
> ```

by changing this to:
```
while (boost::chrono::steady_clock::now() < end_tp && (scan_hist_.size() == 0 || scan_hist_.rbegin()->header.stamp < req.end))
```


<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (changes requested):
Please remove the last sentence from the commit message and make it something like, "This commit should be squashed with the previous 'add service assemble_scans2_blocking' if it is ever upstreamed"

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1#-:-MMb8yGB6YLoiflT3RrT:b-qlac3f)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bobhenz-jabil)



<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1)* status: all files reviewed, 1 unresolved discussion (waiting on @bobhenz-jabil)

*[a discussion](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1#-MMb9Jl5APrk7prfXIy3:-MMb9Jl5APrk7prfXIy4:bs1g8rp) (no related file):*
Please remove the last sentence from the commit message and make it something like, "This commit should be squashed with the previous 'add service assemble_scans2_blocking' if it is ever upstreamed"

---


<!-- Sent from Reviewable.io -->

bobhenz-jabil committed:
Fix crash of laser_assembler if service called before receiving scans

Although this is extremely unlikely to hit on an actual robot, this code
would SEGFAULT if the service was called before any scans had been
received.

I ran into this when trying to run the tilt lidar nodes against a bag file.
In that case it is very easy to hit this problem.

This commit should be squashed with the previous 'add service
assemble_scans2_blocking' if it is ever upstreamed.

c-andy-martin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/laser_assembler/1#-:-MMbAYwPBCnVQcEWp9H4:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bobhenz-jabil)



<!-- Sent from Reviewable.io -->
